File: report-generator/src/main/java/eu/cqse/teamscale/report/testwise/model/TestwiseCoverage.java
package eu.cqse.teamscale.report.testwise.model;

import eu.cqse.teamscale.report.testwise.model.builder.TestCoverageBuilder;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

/** Container for coverage produced by multiple tests. */
public class TestwiseCoverage {

	/** A mapping from test ID to {@link TestCoverageBuilder}. */
	private final Map<String, TestCoverageBuilder> tests = new HashMap<>();

	/**
	 * Adds the {@link TestCoverageBuilder} to the map.
	 * If there is already a test with the same ID the coverage is merged.
	 */
	public void add(TestCoverageBuilder coverage) {
		if (coverage == null || coverage.isEmpty()) {
			return;
		}
		if (tests.containsKey(coverage.getUniformPath())) {
			TestCoverageBuilder testCoverage = tests.get(coverage.getUniformPath());
			testCoverage.addAll(coverage.getFiles());
		} else {
			tests.put(coverage.getUniformPath(), coverage);
		}
	}

	/**
	 * Merges the given {@link TestwiseCoverage} with this one.
	 */
	public void add(TestwiseCoverage testwiseCoverage) {
		if (testwiseCoverage == null) {
			return;
		}
		for (TestCoverageBuilder value : testwiseCoverage.tests.values()) {
			this.add(value);
		}
	}

	public Collection<TestCoverageBuilder> getTests() {
		return tests.values();
	}
}

File: report-generator/src/main/java/com/teamscale/report/testwise/model/TestwiseCoverage.kt
package com.teamscale.report.testwise.model

import com.teamscale.report.testwise.model.builder.TestCoverageBuilder
import java.util.HashMap

/** Container for coverage produced by multiple tests.  */
class TestwiseCoverage {

    /** A mapping from test ID to [TestCoverageBuilder].  */
    private val tests = HashMap<String, TestCoverageBuilder>()

    /**
     * Adds the [TestCoverageBuilder] to the map.
     * If there is already a test with the same ID the coverage is merged.
     */
    fun add(coverage: TestCoverageBuilder?) {
        if (coverage == null || coverage.isEmpty) {
            return
        }
        if (tests.containsKey(coverage.uniformPath)) {
            val testCoverage = tests[coverage.uniformPath]
            testCoverage!!.addAll(coverage.files)
        } else {
            tests[coverage.uniformPath] = coverage
        }
    }

    /**
     * Merges the given [TestwiseCoverage] with this one.
     */
    fun add(testwiseCoverage: TestwiseCoverage?) {
        if (testwiseCoverage == null) {
            return
        }
        for (value in testwiseCoverage.tests.values) {
            this.add(value)
        }
    }

    fun getTests(): Collection<TestCoverageBuilder> {
        return tests.values
    }
}
